src/main/java/com/android/volley/toolbox/BasicAsyncNetwork.java

```diff
@@ -0,0 +1,235 @@
+        }
+        final long requestStartMs = SystemClock.elapsedRealtime();
+        // Gather headers.
+        final Map<String, String> additionalRequestHeaders =
+                HttpHeaderParser.getCacheHeaders(request.getCacheEntry());
+        mAsyncStack.executeRequest(
+                request,
+                additionalRequestHeaders,
+                new AsyncHttpStack.OnRequestComplete() {
+                    @Override
+                    public void onSuccess(HttpResponse httpResponse) {
+                        onRequestSucceeded(request, requestStartMs, httpResponse, callback);
+                    }
+
+                    @Override
+                    public void onAuthError(AuthFailureError authFailureError) {
+                        callback.onError(authFailureError);
+                    }
+
+                    @Override
+                    public void onError(IOException ioException) {
+                        onRequestFailed(
+                                request,
+                                callback,
+                                ioException,
+                                requestStartMs,
+                                /* httpResponse= */ null,
+                                /* responseContents= */ null);
+                    }
+                });
+    }
+
+    /* Helper method that determines what to do after byte[] is received */
+    private void onResponseRead(
+            long requestStartMs,
+            int statusCode,
+            HttpResponse httpResponse,
+            Request<?> request,
+            OnRequestComplete callback,
+            List<Header> responseHeaders,
+            byte[] responseContents) {
+        // if the request is slow, log it.
+        long requestLifetime = SystemClock.elapsedRealtime() - requestStartMs;
+        logSlowRequests(requestLifetime, request, responseContents, statusCode);
+
+        if (statusCode < 200 || statusCode > 299) {
+            onRequestFailed(
+                    request,
+                    callback,
+                    new IOException(),
+                    requestStartMs,
+                    httpResponse,
+                    responseContents);
+        }
+
+        callback.onSuccess(
+                new NetworkResponse(
+                        statusCode,
+                        responseContents,
+                        /* notModified= */ false,
+                        SystemClock.elapsedRealtime() - requestStartMs,
+                        responseHeaders));
+    }
+
+    public static class Builder {
+        private static final int DEFAULT_POOL_SIZE = 4096;
+        AsyncHttpStack mAsyncStack;
+        ByteArrayPool mPool;
+
+        public Builder(AsyncHttpStack httpStack) {
+            mAsyncStack = httpStack;
+            mPool = null;
+        }
+
+        public Builder setPool(ByteArrayPool pool) {
+            mPool = pool;
+            return this;
+        }
+
+        public BasicAsyncNetwork build() {
```

Comment:
throw is mAsyncStack is null?

Reply:
is it better to throw, or add the @NonNull annotation? I get a warning if I do both.

Reply:
OK - we haven't used `@NonNull` much before, but I suppose it's best here because it can be caught at compile time, whereas throwing is only caught at runtime. I'm not 100% convinced b/c it depends on Android Studio tooling and other developers might use other tools that don't detect the problem, but it'll crash at runtime anyway.